components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -51,13 +69,191 @@ public class GoogleOAuth2Authenticator extends OpenIDConnectAuthenticator {
+            }
+        }
+        if (idToken != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("JWT token validated successfully for Google One Tap.");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * This function validates the CSRF double-sided cookie returned from Google One Tap respond
+     * The request is considered as non-attacked request if the CSRF cookie and the parameter is equal.
+     *
+     * @param request Authentication request with Google One Tap auth payloads.
+     * @return Integrity of the authentication request sent via Google One Tap.
+     */
+    private boolean validateCSRFCookies(HttpServletRequest request) {
+
+        if (request.getCookies() == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("No valid cookie found for Google One Tap authentication.");
+            }
+            return false;
+        }
+        List<Cookie> crossRefCookies = Arrays.stream(request.getCookies())
+                .filter(cookie -> cookie.getName().equalsIgnoreCase(G_CSRF_TOKEN))
+                .collect(Collectors.toList());
+
+        if (crossRefCookies.isEmpty() || crossRefCookies.get(0) == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("No CSRF cookie found. Invalid request.");
+            }
+            return false;
+        }
+        String crossRefCookieHalf = crossRefCookies.get(0).getValue();
+        String crossRefParamHalf = request.getParameter(G_CSRF_TOKEN);
+
+        if (StringUtils.isEmpty(crossRefParamHalf) || StringUtils.isEmpty(crossRefCookieHalf)) {
+            if (log.isDebugEnabled()) {
+                log.debug("No CSRF parameter found. Invalid request.");
+            }
+            return false;
+        }
+        if (!crossRefParamHalf.equals(crossRefCookieHalf)) {
+            if (log.isDebugEnabled()) {
+                log.debug("CSRF validation failed for Google One Tap.");
+            }
+            return false;
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Validated CSRF cookies successfully for Google One Tap.");
+        }
+        return true;
+    }
+
+    @Override
+    protected String mapIdToken(AuthenticationContext context, HttpServletRequest request,
+                                OAuthClientResponse oAuthResponse) {
+
+        if (isOneTapEnabled(request)) {
+            // The existence and the validity of CREDENTIAL has been verified at canHandle().
+            String idToken = request.getParameter(CREDENTIAL);
+            context.setProperty(OIDCAuthenticatorConstants.ID_TOKEN, idToken);
+            return idToken;
+        }
+        return super.mapIdToken(context, request, oAuthResponse);
+    }
+
+    @Override
+    protected boolean isInitialRequest(AuthenticationContext context, HttpServletRequest request) {
+        // Google One Tap flow returns the JWT token at the very first callback.
```

Comment:
Shall we add a newline prior to this line?

Reply:
Shall we fix this in other places as well?

Reply:
Fixed f806cee9c82a8896103acc5729765b340d3190c1